astropy/io/fits/src/compressionmodule.c

```diff
@@ -233,159 +235,132 @@ int compress_type_from_string(char* zcmptype) {
 }
 
 
+PyObject *
+get_header_value(PyObject* header, char* key) {
+    PyObject* hdrkey;
+    PyObject* hdrval;
+    hdrkey = PyUnicode_FromString(key);
+    if (hdrkey == NULL) {
+        return NULL;
+    }
+    hdrval = PyObject_GetItem(header, hdrkey);
+    Py_DECREF(hdrkey);
+    PyErr_Clear();
+    return hdrval;
+}
+
+
 // TODO: It might be possible to simplify these further by making the
 // conversion function (eg. PyString_AsString) an argument to a macro or
 // something, but I'm not sure yet how easy it is to generalize the error
 // handling
+/* The get_header_* functions resemble "Header.get" where "def" is the default
+   value, "keyword" is a string representing the header-key and the result is
+   stored in "val".
+   The function returns 0 on success, 1 if the header didn't have the keyword
+   and the default was applied and -1 (with an exception set) if an Exception
+   happened (like a MemoryError or Overflow).
+*/
 int get_header_string(PyObject* header, char* keyword, char* val, char* def) {
-    PyObject* keystr;
-    PyObject* keyval;
-    PyObject* tmp;  // Temp pointer to decoded bytes object
-    int retval;
-
-    keystr = PyUnicode_FromString(keyword);
-    keyval = PyObject_GetItem(header, keystr);
-
-    if (keyval != NULL) {
-        // FITS header values should always be ASCII, but Latin1 is on the
-        // safe side
-        tmp = PyUnicode_AsLatin1String(keyval);
-        strncpy(val, PyBytes_AsString(tmp), 72);
-        Py_DECREF(tmp);
-        retval = 0;
-    }
-    else {
-        PyErr_Clear();
+    PyObject* keyval = get_header_value(header, keyword);
+
+    if (keyval == NULL) {
         strncpy(val, def, 72);
-        retval = 1;
+        return PyErr_Occurred() ? -1 : 1;
     }
-
-    Py_DECREF(keystr);
-    Py_XDECREF(keyval);
-    return retval;
+    PyObject* tmp = PyUnicode_AsLatin1String(keyval);
+    // FITS header values should always be ASCII, but Latin1 is on the
+    // safe side
+    Py_DECREF(keyval);
+    if (tmp == NULL) {
+        /* could always fail to allocate the memory or such like. */
+        return -1;
+    }
+    strncpy(val, PyBytes_AsString(tmp), 72);
+    Py_DECREF(tmp);
+    return 0;
 }
 
 
-int get_header_int(PyObject* header, char* keyword, int* val, int def) {
-    PyObject* keystr;
-    PyObject* keyval;
-    int retval;
-
-    keystr = PyUnicode_FromString(keyword);
-    keyval = PyObject_GetItem(header, keystr);
+int get_header_long(PyObject* header, char* keyword, long* val, long def) {
```

Comment:
This is getting really pedantic, but technically the function signatures for this and similar functions should
```c
int get_header_long(const PyObject* header, const char* keyword, long* val, long def)
```

Reply:
Hm, why not `(const PyObject * const header, const char* const keyword, long* val, const long def)`? I mean we don't change the pointers and not the values pointed to and `def` is technically also `const`. I haven't used `const`s much so let me know if I'm wrong.

Reply:
Good question: the point of using `const` this way is to guarantee to callers that certain data that you pass in to the function will not be unexpectedly changed by the function. So while the example you've given is not wrong, it's also not necessary. 

The value of `def` can be changed by the function, but any changes to it will only be visible within the scope of the function, so there's no need to add a `const`. Similarly, the pointers to `header` and `keyword` could be changed within the function, but those changes won't be visible outside the function.

However, the function **could** theoretically change the _contents_ of `header` and `keyword`, and those changes **would** be visible to the caller. So we add `const` to those parameters to promise that the contents won't be changed by the function.

We do of course expect the contents of `val` to be updated by the function, so we do not use `const` there.

Reply:
I thought `const` is just used at compile-time and this is a purely internal function so just useful for the compiler warnings/errors thrown. Or am I wrong?

I can change it according to your suggestion, I just (personally) don't see the point of `const` here. Just let me know what you prefer, I'll change it :)

Reply:
But thank you for the explanation. 

I never bothered much with the rationale of `const` so this is extremely useful advise!

Reply:
Everything in C is just used at compile time 😛 . But seriously, the compiler can help you find potential bugs before your code even runs. That's amazing! Using `const` in this way lets the compiler determine whether you have inadvertently modified a buffer that should have been read-only.

You have a point that this is more important in public-facing API functions than it is in an internal function like this, but adding `const` correctness here can still help you prevent buggy code that would be otherwise be pretty hard to diagnose.

I'm not going to press for it to be changed here, but it is good to keep in mind for future work.

Reply:
okay, I added the `const char *` ones, but I didn't do `const PyObject * header`. That would throw unnecessary warnings with `PyObject_GetItem` because that cannot garantuee `const`.